### DISCUSSION TITLE
Home Assistant: Utility Meter "jumps" up & down after server reboot

### DISCUSSION THREAD

**Discussion by meironke** [jomjol/AI-on-the-edge-device#2009] [General]
### The Problem

Thank you for this awesome project. Integration works perfectly and helps us reducing water consumption 🙂

### My issue
**Each night (after restarting my server incl. mqtt mosquitto & Home Assistant) the water value jumps up and down again.**

Here's how it looks like:
![2023-02-10 12_53_57-KNX – Home Assistant](https://user-images.githubusercontent.com/56265223/218089306-7cf9cb84-ad1b-412d-8f05-60884ec3c562.png)

Do you have any ideas why and where this issue is happening and how I can resolve this?

### My setup
- Home Assistant (HA) & mqtt services all setup within a Docker environment
- Integration of the watermeter into HA via mqtt discovery
- Utiliy meter is calculating the daily usage (config see below)
- Mosquitto config is set to `persistence true`
- Retain flag is set to `true`

### Utility meter configuration
```
water_daily:
  source: sensor.watermeter_value
  name: "Wasserverbrauch (Tag)"
  cycle: daily
```

### Version

Release: v13.0.8 (Commit: 04f2f23+)

### Logfile

```shell
2023-02-09T21:42:35+0100,main,00550.8151,550.8151,550.8151,0.000000,0.0000,no error,9.9,9.9,5.0,5.1,9.9,7.9,1.4,5.4,1.7
2023-02-09T21:52:35+0100,main,00550.8154,550.8154,550.8154,0.000030,0.0003,no error,10.0,9.9,5.0,5.2,9.9,7.9,1.5,5.6,4.1
2023-02-09T22:02:35+0100,main,00550.8154,550.8154,550.8154,0.000000,0.0000,no error,9.9,10.0,5.0,5.1,9.9,7.9,1.5,5.6,4.1
2023-02-09T22:12:35+0100,main,00550.8154,550.8154,550.8154,0.000000,0.0000,no error,10.0,10.0,5.0,5.1,9.9,7.9,1.5,5.6,4.1
2023-02-09T22:22:35+0100,main,00550.8154,550.8154,550.8154,0.000000,0.0000,no error,10.0,9.9,5.0,5.1,9.9,7.9,1.5,5.6,4.1
2023-02-09T22:32:35+0100,main,00550.8154,550.8154,550.8154,0.000000,0.0000,no error,10.0,9.9,5.0,5.1,9.9,7.9,1.4,5.6,4.1
2023-02-09T22:42:35+0100,main,00550.8154,550.8154,550.8154,0.000000,0.0000,no error,10.0,9.9,5.0,5.1,9.9,7.9,1.4,5.6,4.1
2023-02-09T22:52:35+0100,main,00550.8154,550.8154,550.8154,0.000000,0.0000,no error,10.0,9.9,5.0,5.1,9.8,7.9,1.5,5.6,4.1
2023-02-09T23:02:35+0100,main,00550.8154,550.8154,550.8154,0.000000,0.0000,no error,9.9,9.9,5.0,5.1,9.9,7.9,1.5,5.6,4.1
2023-02-09T23:12:35+0100,main,00550.8184,550.8184,550.8184,0.000300,0.0030,no error,10.0,9.9,5.0,5.1,9.9,7.9,1.7,8.7,4.0
2023-02-09T23:22:34+0100,main,00550.8286,550.8286,550.8286,0.001022,0.0102,no error,10.0,9.9,5.0,5.1,9.9,8.0,2.7,8.6,6.4
2023-02-09T23:32:35+0100,main,00550.8295,550.8295,550.8295,0.000090,0.0009,no error,10.0,9.9,5.0,5.1,9.9,8.0,2.8,9.4,5.3
2023-02-09T23:42:34+0100,main,00550.8295,550.8295,550.8295,0.000000,0.0000,no error,10.0,10.0,5.0,5.1,9.9,8.0,2.8,9.4,5.3
2023-02-09T23:52:34+0100,main,00550.8295,550.8295,550.8295,0.000000,0.0000,no error,9.9,10.0,5.0,5.1,9.9,8.0,2.8,9.3,5.3
```

### Expected Behavior

Ths utility meter should not jump up and back down again but should remain stable.

**Comment by supaeasy** [comment]
Afaik Daily Cycle means resetting the value. So I don't know why it goes up but why it goes down. I am guessing: Do you restart your server shortly after 00:00 or before? If before: "Retain" flag _might_ lead to adding up yesterdays value before resetting the counter via daily cycle.

Three questions: 
- why do you restart your server daily? Is there a special reason? That should not be necessary
- does the value always jump up to or by the exact same value? Or is it yesterdays value x2?
- Why don't you use the energy board? The visualisation is way prettier.

**Comment by meironke** [comment]
Thank you for your reply 🙂 

Regarding your questions.
- I am rebooting to update the system (packages, host, etc.) Scheduled daily, weekly or monthly does not matter, behaviour is the same.
- Reboot is (now) at 11:35pm, system should be up and running latest at 11:40pm, and with a 10min interval for the ESP module the new value should be in latest 11:50pm (before midnight). Had this previously at 2:00am, but this ruined the value for the day.
- The value is jumping random. See the graph attached.
- Thanks for the hint with the energy board. Wasn't aware - that's quite new, isn't it? I'll give it a shot and hope, that calculation is stable.

![image](https://user-images.githubusercontent.com/56265223/218147011-39870746-f7cb-4749-8758-41aa6895c03f.png)

**Reply by caco3** [reply]
I had myself problems with the Utility Meter, see https://github.com/home-assistant/frontend/issues/13995 here nobody from Homeassistant was willing to look into it. I never got it working the way it should. But since some months, Homeassistant supports also water consumption  in the Energy Dashboard. That works without any issues for me:
![grafik](https://user-images.githubusercontent.com/1783586/218190313-98869787-4e51-4c0a-ad4f-a4e3d1561922.png)

I suggest you also try it this way.

**Comment by yuqianma** [comment]
‌‌I also experienced a value jump after HA reboot. The value is always a stale value because the values are the same and uptime is "93"s. The board has been up for at least days. I don't know where it's from.

<img width="477" alt="image" src="https://github.com/jomjol/AI-on-the-edge-device/assets/3090650/76fe673a-dfd8-4368-8f14-d56c89baea00">

Then I tried:

1. My MQTT setting 'retain' was set to false. I attempted to change it to true. Restarted the ESP, restarted the HA. Still the same outdated value.
2. Turned off polling for the MQTT integration option. Restarted the HA. No impact observed. ( I thought it might have polled wrong value. )
3. Turned off the ESP32-Cam board, restarted the HA. Turn on the board. No outdated value! Only the last value! ( It should be the expected result of "retain"? ) Even after I reboot the HA again, the value didn't jump.

Though strange, the temporary fix for my case is to unplug the board and restart the HA and plug in the board again.

**Reply by caco3** [reply]
To point 1.:
The changed retain flag might only have an impact after the first round got sent! Maybe you did not wait long enough after restart of the ESP32?

**Reply by yuqianma** [reply]
The first round sent right after the board boot, right?
So even though it might be the case in the first HA restart, the value was still wrong in the second HA restart when I changed the polling option.

But I'll take note of it if I come across the issue again.

**Reply by caco3** [reply]
the first round starts shortly after powerup, but it only sends the new values at its end, so it can take some minutes!

**Reply by caco3** [reply]
Please show a screenshot of your Data Graph (Menu `Data > Data Graph`). If it is the device who does the error, you would see it there as well.
But I strongly believe it is an issue with the HA Utility Meter.
I had issues myself to get it working reliably, see https://jomjol.github.io/AI-on-the-edge-device-docs/Integration-Home-Assistant/#statistics-graph. However since HA supports Engergy Devices, this no longer is needed!

**Reply by yuqianma** [reply]
Since it's been a while now, I can't guarantee the order in which the data were sent. And I turned off data logging after the SD card broke once to reduce data writing. So because my memory may not be reliable, anything is possible.
My guess is that some MQTT mechanism or issues caused a stagnant state. But it's quite stable these days, I don't really want to reproduce this issue either.
Thanks!
